File: javascript/js/FileHandler.js
class FileHandler {
  constructor(onFileLoaded) {
    this.onFileLoaded = onFileLoaded;
    this.dropZone   = document.getElementById('drop-zone');
    this.fileInput  = document.getElementById('file-input');
    this.dropLabel  = document.getElementById('drop-label');
    this._bindEvents();
  }

  _bindEvents() {
    this.dropZone.addEventListener('dragover', e => {
      e.preventDefault();
      this.dropZone.classList.add('drag-over');
    });
    this.dropZone.addEventListener('dragleave', () => {
      this.dropZone.classList.remove('drag-over');
    });
    this.dropZone.addEventListener('drop', e => {
      e.preventDefault();
      this.dropZone.classList.remove('drag-over');
      if (e.dataTransfer.files[0]) this.handleFile(e.dataTransfer.files[0]);
    });
    this.fileInput.addEventListener('change', e => {
      if (e.target.files[0]) this.handleFile(e.target.files[0]);
    });
  }

  handleFile(file) {
    const reader = new FileReader();
    reader.onload = e => {
      try {
        let rows, sheetName = '';
        if (file.name.endsWith('.csv')) {
          const wb = XLSX.read(e.target.result, { type: 'binary' });
          sheetName = wb.SheetNames[0];
          rows = XLSX.utils.sheet_to_json(wb.Sheets[sheetName], { defval: '' });
        } else {
          const wb = XLSX.read(new Uint8Array(e.target.result), { type: 'array' });
          sheetName = wb.SheetNames[0];
          rows = XLSX.utils.sheet_to_json(wb.Sheets[sheetName], { defval: '' });
        }

        if (!rows.length) throw new Error('File appears empty or unreadable.');

        const columns     = Object.keys(rows[0]);
        const dataContext = this.buildDataContext(rows, columns, file.name, sheetName);

        this.dropZone.classList.add('loaded');
        this.dropLabel.className   = 'drop-label loaded';
        this.dropLabel.textContent = `✓ ${file.name}`;

        this.onFileLoaded({ rows, columns, sheetName, fileName: file.name, dataContext });
      } catch (err) {
        this.onFileLoaded(null, err.message);
      }
    };
    if (file.name.endsWith('.csv')) reader.readAsBinaryString(file);
    else reader.readAsArrayBuffer(file);
  }

  sanitize(val) {
    return String(val)
      .substring(0, 60)
      .replace(/\\/g, '')
      .replace(/"/g, "'")
      .replace(/[\r\n\t]/g, ' ')
      .replace(/[^\x20-\x7E]/g, '')
      .trim();
  }

  buildDataContext(rows, cols, fileName, sheetName) {
    const totalRows = rows.length;

    const colStats = cols.map(col => {
      const vals = rows.map(r => r[col]).filter(v => v !== '' && v != null);
      const nums = vals.filter(v => !isNaN(Number(v)) && v !== '');
      const isNumeric = nums.length / vals.length > 0.8;
      if (isNumeric && nums.length) {
        const ns  = nums.map(Number);
        const sum = ns.reduce((a, b) => a + b, 0);
        const min = Math.min(...ns);
        const max = Math.max(...ns);
        const avg = (sum / ns.length).toFixed(2);
        return `${this.sanitize(col)} [numeric]: min=${min}, max=${max}, avg=${avg}, sum=${sum.toFixed(2)}, count=${ns.length}`;
      } else {
        const unique  = [...new Set(vals)];
        const preview = unique.slice(0, 8).map(v => String(v).substring(0, 30)).join(', ');
        return `${this.sanitize(col)} [text]: ${unique.length} unique values — e.g. ${preview}${unique.length > 8 ? '…' : ''}`;
      }
    });

    const sampleRows = totalRows <= 30 ? rows : [...rows.slice(0, 20), ...rows.slice(-5)];
    const sampleCsv  = [
      cols.map(c => this.sanitize(c)).join(' | '),
      ...sampleRows.map(r => cols.map(c => this.sanitize(r[c]).substring(0, 40)).join(' | '))
    ].join('\n');

    return `
FILE: ${fileName} (Sheet: "${sheetName}")
TOTAL ROWS: ${totalRows.toLocaleString()} | COLUMNS: ${cols.length}

COLUMN STATISTICS:
${colStats.join('\n')}

SAMPLE DATA (${sampleRows.length} rows shown):
${sampleCsv}
`.trim();
  }
}

File: python/src/file_handler.py
import io
import re
import pandas as pd


class FileHandler:

    FULL_ROW_LIMIT = 30   # Send all rows below this
    SAMPLE_SIZE    = 30   # If above limit, sample this many evenly

    def parse(self, file_bytes: bytes, filename: str, field_context: str = "") -> dict:
        """Parse an uploaded Excel or CSV file and return structured data."""
        if filename.endswith(".csv"):
            df = pd.read_csv(io.BytesIO(file_bytes))
            sheet_name = "Sheet1"
        else:
            xf = pd.ExcelFile(io.BytesIO(file_bytes))
            sheet_name = xf.sheet_names[0]
            df = pd.read_excel(io.BytesIO(file_bytes), sheet_name=sheet_name)

        df = df.fillna("")
        rows    = df.to_dict(orient="records")
        columns = list(df.columns)

        return {
            "rows":         rows,
            "columns":      columns,
            "sheet_name":   sheet_name,
            "file_name":    filename,
            "data_context": self.build_data_context(df, columns, filename, sheet_name, field_context)
        }

    @staticmethod
    def sanitize(val) -> str:
        val = str(val)[:60]
        val = re.sub(r'[\\"\r\n\t]', ' ', val)
        val = re.sub(r'[^\x20-\x7E]', '', val)
        return val.strip()

    def build_data_context(self, df: pd.DataFrame, columns: list, file_name: str, sheet_name: str, field_context: str = "") -> str:
        total_rows = len(df)
        col_stats  = []

        for col in columns:
            series     = df[col].replace("", pd.NA).dropna()
            if len(series) == 0:
                continue
            numeric    = pd.to_numeric(series, errors="coerce").dropna()
            is_numeric = len(numeric) / len(series) > 0.8

            if is_numeric and len(numeric):
                col_stats.append(
                    f"{self.sanitize(col)} [numeric]: "
                    f"min={numeric.min()}, max={numeric.max()}, "
                    f"avg={numeric.mean():.2f}, sum={numeric.sum():.2f}, "
                    f"count={len(numeric)}"
                )
            else:
                unique  = series.unique()
                preview = ", ".join(str(v)[:30] for v in unique[:8])
                ellipsis = "..." if len(unique) > 8 else ""
                col_stats.append(
                    f"{self.sanitize(col)} [text]: {len(unique)} unique values - e.g. {preview}{ellipsis}"
                )

        # Smart row limit — all rows under threshold, even sample above
        if total_rows <= self.FULL_ROW_LIMIT:
            data_df  = df
            data_label = f"DATA (all {total_rows:,} rows)"
        else:
            step     = total_rows // self.SAMPLE_SIZE
            data_df  = df.iloc[::step].head(self.SAMPLE_SIZE)
            data_label = f"DATA (evenly sampled {len(data_df)} of {total_rows:,} rows)"

        header_row = " | ".join(self.sanitize(c) for c in columns)
        data_rows  = "\n".join(
            " | ".join(self.sanitize(str(row[col]))[:40] for col in columns)
            for _, row in data_df.iterrows()
        )

        parts = [
            f"FILE: {file_name} (Sheet: \"{sheet_name}\")",
            f"TOTAL ROWS: {total_rows:,} | COLUMNS: {len(columns)}",
        ]

        # Inject field context from markdown if provided
        if field_context and field_context.strip():
            parts += [
                "",
                "FIELD CONTEXT & BUSINESS RULES:",
                field_context.strip(),
            ]

        parts += [
            "",
            "COLUMN STATISTICS:",
            "\n".join(col_stats),
            "",
            f"{data_label}:",
            header_row,
            data_rows,
        ]

        return "\n".join(parts).strip()